tools/data-handler/src/resources/resource-object.ts

```diff
@@ -0,0 +1,100 @@
+/**
+    Cyberismo
+    Copyright © Cyberismo Ltd and contributors 2024
+
+    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License version 3 as published by the Free Software Foundation.
+
+    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public
+    License along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
+
+import { Project } from '../containers/project.js';
+import { ResourceName } from '../utils/resource-utils.js';
+import {
+  FileBasedResources,
+  FolderBasedResources,
+} from '../interfaces/resource-interfaces.js';
+
+/**
+ * Abstract class for resources.
+ *
+ */
+export abstract class AbstractResource {
```

Comment:
To clarify. You mean "show all cardTypes" from cardType resource class (or its base class)?
And similarly for all other resource types. Or single "show all resources in the project" API?

The integration to `ResourceCollector` is not yet there. I need to think how it would be implemented properly.

Reply:
Yes, by show all I meant that we have a way of getting all resources of a certain type. But my point was that it would be better if it happens inside these resource classes, so that it abstracts the file system without having to do ifs in `ResourceCollector`. But they cannot be class methods, since then you'd need to create an instance of a resource before getting all resources. I guess that's why repository pattern is pretty common, but I **don't think** that's good for this since our resources are so similar that it will just create additional boilerplate
![image](https://github.com/user-attachments/assets/82f7132d-2538-4eb5-a0f8-ad915bf08b33)

Reply:
I think I will keep the `ResourceCollector` as the "collection" of resources. These (resource classes) are individual instances of one resource; here instance is the data of the metadata JSON + how it handles actions. The classes handle "actions";  what happens to a resource when certain command is applied on it. 
They are not there to collect resource types for others. 
Whereas resource collection is only interested in updates (add, remove, change name), or rebuild the whole collection. 

Logically, all of our resources are the same, but they have some business logic built into them (e.g. if workflow name changes, card types need to be updated).

What is now lacking; is that the two -`ResourceCollector` and resource classes -  are not that well integrated yet. I'd like to move ALL code that handles (read, write, delete) metadata JSON files into the resource classes. 
Collector would just scan the filesystem; not read file contents.